src/api/XRSession.js

```diff
@@ -46,15 +46,114 @@ export default class XRSession extends EventTarget {
       outputContext,
       ended: false,
       suspended: false,
-      suspendedCallback: null,
+      frameCallbacks: [],
+      frameHandle: 0,
+      deviceFrameHandle: null,
       id,
       activeRenderState: new XRRenderState(),
       pendingRenderState: null,
       currentInputSources: []
     };
 
-    const frame = new XRFrame(device, this, this[PRIVATE].id);
-    this[PRIVATE].frame = frame;
+    // Single handler for animation frames from the device. The spec says this must
+    // run on every candidate frame even if there are no callbacks queued up.
+    this[PRIVATE].onDeviceFrame = () => {
+      if (this[PRIVATE].ended) {
+        return;
+      }
+
+      // - If session’s pending render state is not null, apply the pending render state.
+      if (this[PRIVATE].pendingRenderState !== null) {
+        this[PRIVATE].activeRenderState = new XRRenderState(this[PRIVATE].pendingRenderState);
+        this[PRIVATE].pendingRenderState = null;
+
+        // TODO: set compositionDisabled
+
+        // Report to the device since it'll need to handle the layer for rendering.
+        if (this[PRIVATE].activeRenderState.baseLayer) {
+          this[PRIVATE].device.onBaseLayerSet(
+            this[PRIVATE].id,
+            this[PRIVATE].activeRenderState.baseLayer);
+        }
+
+        if (this[PRIVATE].activeRenderState.inlineVerticalFieldOfView) {
+          this[PRIVATE].device.onInlineVerticalFieldOfViewSet(
+            this[PRIVATE].id,
+            this[PRIVATE].activeRenderState.inlineVerticalFieldOfView);
+        }
+      }
+
+      // - If session’s renderState's baseLayer is null, abort these steps.
+      if (this[PRIVATE].activeRenderState.baseLayer === null) {
+        return;
```

Comment:
Just returning will halt the frame loop, since you don't queue up the next frame. Since I don't think is what you want lets either always queue up the next frame earlier in the function or do queue up the next one explicitly before returning here.

Reply:
Done by moving the queuing up.